### PR TITLE
Adds from_random_bytes to curves and makes from_random_bytes match deserialization

### PR DESCRIPTION
**Pull request by kobigurk**
This is done by introducing `from_random_bytes_with_flags` to fields.

### CODE REVIEW

File: algebra-core/src/curves/models/short_weierstrass_jacobian.rs [hunk 1]
```diff
@@ -110,6 +111,17 @@ impl<P: Parameters> GroupAffine<P> {
...
             .is_zero()
     }
+
+    pub fn from_random_bytes(bytes: &[u8]) -> Option<Self> {
+        P::BaseField::from_random_bytes_with_flags(bytes).and_then(|(x, flags)| {
+            let parsed_flags = SWFlags::from_u8(flags);
+            if parsed_flags.is_infinity() {
+                Some(Self::zero())
+            } else {
+                Self::get_point_from_x(x, parsed_flags.is_positive().unwrap())
```
Comment (gakonst) [inline algebra-core/src/curves/models/short_weierstrass_jacobian.rs]: Can we handle the unwrap?

File: algebra-core/src/curves/models/short_weierstrass_projective.rs [hunk 2]
```diff
@@ -112,6 +113,17 @@ impl<P: Parameters> GroupAffine<P> {
...
             .is_zero()
     }
+
+    pub fn from_random_bytes(bytes: &[u8]) -> Option<Self> {
+        P::BaseField::from_random_bytes_with_flags(bytes).and_then(|(x, flags)| {
+            let parsed_flags = SWFlags::from_u8(flags);
+            if parsed_flags.is_infinity() {
+                Some(Self::zero())
+            } else {
+                Self::get_point_from_x(x, parsed_flags.is_positive().unwrap())
```
Comment (gakonst) [inline algebra-core/src/curves/models/short_weierstrass_projective.rs]: Can we handle the unwrap?

File: algebra-core/src/fields/macros.rs [hunk 3]
```diff
@@ -274,7 +281,18 @@ macro_rules! impl_Fp {
         impl<P: $FpParameters> FromBytes for $Fp<P> {
             #[inline]
             fn read<R: Read>(reader: R) -> IoResult<Self> {
-                BigInteger::read(reader).map($Fp::from_repr)
+                BigInteger::read(reader).and_then( |b|
+                    if b.is_zero() {
+                        Ok($Fp::zero())
+                    } else {
+                        let f = $Fp::from_repr(b);
+                        if f == $Fp::zero() {
```
Comment (gakonst) [inline algebra-core/src/fields/macros.rs:289]: This is nice, it checks if the BigNumber is 0, and returns Fp::zero, and in the else branch, if the Fp::from_repr returned zero, it means that there was an error during reading

**Review by gakonst — APPROVED** [review]

File: algebra-core/src/curves/models/short_weierstrass_projective.rs (hunk 2 above)
Comment (kobigurk) [reply algebra-core/src/curves/models/short_weierstrass_projective.rs]: Done

File: algebra-core/src/fields/macros.rs (hunk 3 above)
Comment (kobigurk) [reply algebra-core/src/fields/macros.rs:289]: Yep! Should definitely think about improving `from_repr` though

**Pull request merged**